wdl-analysis/src/types/v1.rs

```diff
@@ -692,6 +693,33 @@ impl<'a, C: EvaluationContext> ExprTypeEvaluator<'a, C> {
         self.placeholders -= 1;
     }
 
+    fn invalid_placeholder_option(
+        ty: &Type,
+        _span: Span,
+        option: Option<PlaceholderOption>,
+    ) -> Diagnostic {
+        let message = match option {
+            Some(PlaceholderOption::Sep(_)) => format!(
+                "type mismatch for placeholder option `sep`: expected type `Array[P]` where P: any primitive type or Union, but found `{}`",
+                ty
```

Comment:
Nit: I generally prefer we put the expressions in the placeholder, makes the formatted string easier to read in source:

```suggestion
                "type mismatch for placeholder option `sep`: expected type `Array[P]` where P: any primitive type, but found `{ty}`"
```

We should also not have `Union` in any error messages; it's a hidden type and generally unknown to the users of WDL.